### ISSUE TITLE
Replace async module usage by ES6 Promises

### ISSUE THREAD

**Issue by fmcarvalho** [sahat/hackathon-starter#633]
I removed the `async` package from _hackathon-starter_ and replace it by es6 `Promises` use of `then()` and `all()`, which behave like `async::waterfall()` and `async::parallel()`. **This adaptation is available on branch [feature/promises](https://github.com/fmcarvalho/hackathon-starter/tree/feature/promises)**.

For instance check the final result of the [`postForgot()`](https://github.com/fmcarvalho/hackathon-starter/blob/feature/promises/controllers/user.js#L367) method:

```js
return Promise.all([
  createRandomToken(),
  User.findOne({ email: req.body.email }).exec()
])
.then(setRandomToken)
.then(sendForgotPasswordEmail)
.then(() => res.redirect('/forgot'))
.catch(err => next(err));
```

Since mongoose operations (like `.save()` and queries) return Promises/A+ conformant promises, we could take advantage of `then()` to implement the same behavior of `waterfall()`.

Moreover `transporter.sendMail()` also returns a Promise object If no callback argument is set. So, we can chain promises from `mongoose` and `nodemailer`

In [controllers/users.js](https://github.com/fmcarvalho/hackathon-starter/blob/feature/promises/controllers/user.js) I made an adaptation which removes `require(async)` and replaces `waterfall()` by `then()` usage. IMHO the final code is clean and comprehensive because you can clearly follow the sequence among several operations.

What do you think about it? Can I make a pull request of this adaptation?

**Comment by fmcarvalho** [comment]
In truth we do not need the `async` module at all and we can achieve the same behavior of `waterfall()` and `parallel()` through the use of  `Promise::then()` and `Promise::all()`. 

I removed the `async` module from dependencies in `package.json` and replace `async` usage by ES6 `Promise` in [`api.js`](https://github.com/fmcarvalho/hackathon-starter/blob/feature/promises/controllers/api.js) and [`users.js`](https://github.com/fmcarvalho/hackathon-starter/blob/feature/promises/controllers/user.js). 

This adaptation is available on branch [feature/promises](https://github.com/fmcarvalho/hackathon-starter/tree/feature/promises).

**Comment by YasharF** [comment]
+1
I think there might also be a bug in the error handling of the existing async.waterfall implementation.  @fmcarvalho can you check in your branch and see if the following issue exists or if it is/can be fixed?

Currently form what I am experiencing: If at https://github.com/sahat/hackathon-starter/blob/master/controllers/user.js#L289 sendMail fails to send the email, and returns an error (i.e. "Error: Connection closed") then the application crashes with a 500 code, instead of gracefully handling the error with a flash message that the email could not be sent, please contact the developer by email at abc@devteam.mailserver.com.  If my finding is correct about the existing implementation, having the new Promise based implementation handle the error gracefully would be needed as well.

In our case it ends up with the following in the browser of the user:
## **Connect**
_500_ **Error: Connection closed**
   at SMTPConnection.<anonymous> (/workspace/myapp/node_modules/nodemailer-smtp-transport/lib/smtp-transport.js:113:29)
   at SMTPConnection.g (events.js:291:16)
   at emitNone (events.js:86:13)
   at SMTPConnection.emit (events.js:185:7)
   at SMTPConnection._destroy (/workspace/myapp/node_modules/smtp-connection/lib/smtp-connection.js:593:10)
   at SMTPConnection._onEnd (/workspace/myapp/node_modules/smtp-connection/lib/smtp-connection.js:573:10)
   at TLSSocket.g (events.js:291:16)
   at emitNone (events.js:91:20)
   at TLSSocket.emit (events.js:185:7)
   at endReadableNT (_stream_readable.js:974:12)
   at _combinedTickCallback (internal/process/next_tick.js:74:11)
   at process._tickCallback (internal/process/next_tick.js:98:9)

**Comment by fmcarvalho** [comment]
Yes, my implementation  [feature/promises](https://github.com/fmcarvalho/hackathon-starter/tree/feature/promises) produces exactly the same behavior that you described and I think that is the expected behavior.

Maybe that is not a Bug. I think flash messages are used to alert users about client errors, e.g. when the user introduces an invalid email the application notifies him with [“_Account with that email address does not exist_”]( https://github.com/fmcarvalho/hackathon-starter/blob/feature/promises/controllers/user.js#L338). 

In your use case you are describing a server error e.g. "_Error: Connection closed_", and the application `app.js` should handle all server errors in a consistent way. In this case the _hackathon-starter_ `app.js` handles server errors in [`errorHandler() ` middleware]( https://github.com/sahat/hackathon-starter/blob/master/app.js#L220), which sends back to the client full error stack traces. 

This is an application design choice. So, in this case, whenever a server error occurs it just forwards that error to the error middleware through [`next(err)` statement]( https://github.com/sahat/hackathon-starter/blob/master/controllers/user.js#L295). And I keep this same approach doing [`.catch(err => next(err));`]( https://github.com/fmcarvalho/hackathon-starter/blob/feature/promises/controllers/user.js), which in truth could be simplified to `.catch(next);`

Anyway, maybe you are right about the annoying stack trace which ends up in the browser of the user. But IMHO the correct approach to avoid that behavior is to replace the [`app.use(errorHandler());` middleware]( https://github.com/sahat/hackathon-starter/blob/master/app.js#L220) by something that gracefully handles the error: 1) logs that error occurrence in the server and 2) simultaneously sends back to the client a more user friendly message.

**Comment by teogenesmoura** [comment]
I know this thread hasn't been active for a while now, but I think it is a valid doubt:
What is the advantage of taking out async references in favor of promises? Would it be just to remove a dependency from the project or would there be performance gains? I'm guessing it might be related to cleaner code, but since I'm kind of new to javascript I thought it'd be a good idea to ask

**Comment by mikesankovich** [comment]
Concise/more readable code. It reads like a synchronous program, but with the advantages of async.

Better error handling, as you can handle synchronous and asynchronous code in the same ways using try/catch.

**Comment by fmcarvalho** [comment]
This was already merged in pull request #638. There is no more async dependency in package.json right now. So, I think the project owner agrees with all the advantages of using Promises.
